### ISSUE TITLE
Crashes when recovering app from background state on Android

### ISSUE THREAD

**Issue by Aeyoun** [piwik/piwik-mobile-2#5359]
Reproduction steps:

1. Open app by tapping its icon
2. Click the Home button to close the app
3. Start the app again by tapping it's app icon

On a few occasions I need step 2.a as well: "Open a bunch of other apps before proceeding to step 3", but most of the time I can reproduce it right away without this step.

I submitted a crash log through Google Play just now with the comment "ThisIsQuiteSomething". I’ve been submitting logs for a month, so if anyone were actually looking at these it should have come up already

Android 7, latest app versions, Nexus 5X. I had this problem a year ago with Android 6 too on a Xperia Z3. Stopped using the app back then, but thought I’d give it another go.

    10-23 16:13:17.520  4890 28311 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=org.piwik.mobile2/.PiwikMobile2Activity (has extras)} from uid 10031 on display 0
    10-23 16:13:17.639  4890 30075 I ActivityManager: Start proc 24509:org.piwik.mobile2/u0a225 for activity org.piwik.mobile2/org.appcelerator.titanium.TiActivity
    10-23 16:13:17.674 23929 23929 I MemoryTrimmer: Trimming objects from memory, since app is in the background.
    10-23 16:13:17.796 24509 24509 I art     : Starting a blocking GC AddRemoveAppImageSpace
    10-23 16:13:17.850 24509 24509 I TiApplication: (main) [0,0] checkpoint, app created.
    10-23 16:13:17.966 24509 24509 I TiApplication: (main) [117,117] Titanium 3.4.1 (2014/11/13 13:33 5982e8f)
    10-23 16:13:18.157 24509 24524 W linker  : /data/app/org.piwik.mobile2-1/lib/arm/libstlport_shared.so has text relocations. This is wasting memory and prevents security hardening. Please fix.
    10-23 16:13:18.165 24509 24524 W linker  : /data/app/org.piwik.mobile2-1/lib/arm/libkroll-v8.so has text relocations. This is wasting memory and prevents security hardening. Please fix.
    10-23 16:13:18.200 24509 24509 W linker  : /data/app/org.piwik.mobile2-1/lib/arm/libtiverify.so has text relocations. This is wasting memory and prevents security hardening. Please fix.
    10-23 16:13:18.226 24509 24509 I TiApplication: (main) [260,377] Titanium Javascript runtime: v8
    10-23 16:13:18.256 24509 24509 W TiBaseActivity: (main) [30,407] Runtime has been disposed or app has been killed. Finishing.
    10-23 16:13:18.260 24509 24509 W TiApplication: (main) [4,411] Scheduling application restart
    10-23 16:13:18.345 24509 24524 W V8Object: (KrollRuntimeThread) [85,496] Runtime disposed, cannot set property 'userAgent'
    10-23 16:13:18.399  4890 25994 I WindowManager: Failed to capture screenshot of Token{13f69b6 ActivityRecord{90c6d51 u0 org.piwik.mobile2/.PiwikMobile2Activity t4615 f}} appWin=Window{dba7c28 u0 Starting org.piwik.mobile2} drawState=1
    10-23 16:13:18.491  4890 14190 W InputMethodManagerService: Window already focused, ignoring focus gain of: com.android.internal.view.IInputMethodClient$Stub$Proxy@6295379 attribute=null, token = android.os.BinderProxy@5ea6a02

    10-23 16:13:23.659  4890  6470 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x4 cmp=org.piwik.mobile2/.PiwikMobile2Activity (has extras)} from uid 10225 on display 0
    10-23 16:13:23.660  4890  6470 W ActivityManager: startActivity called from non-Activity context; forcing Intent.FLAG_ACTIVITY_NEW_TASK for: Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x4 cmp=org.piwik.mobile2/.PiwikMobile2Activity (has extras) }
    10-23 16:13:23.726 24509 24509 I TiRootActivity: (main) [0,0] checkpoint, on root activity create, savedInstanceState: null
    10-23 16:13:23.728 24509 24509 I TiApplication: (main) [2,2] Analytics have been disabled
    10-23 16:13:23.743 24509 24524 F libc    : Fatal signal 11 (SIGSEGV), code 2, fault addr 0x4df00024 in tid 24524 (KrollRuntimeThr)
    10-23 16:13:23.744   363   363 W         : debuggerd: handling request: pid=24509 uid=10225 gid=10225 tid=24524
    10-23 16:13:23.872 24532 24532 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
    10-23 16:13:23.872 24532 24532 F DEBUG   : Build fingerprint: 'google/bullhead/bullhead:7.1.1/NPF10C/3347772:user/release-keys'
    10-23 16:13:23.872 24532 24532 F DEBUG   : Revision: 'rev_1.0'
    10-23 16:13:23.872 24532 24532 F DEBUG   : ABI: 'arm'
    10-23 16:13:23.872 24532 24532 F DEBUG   : pid: 24509, tid: 24524, name: KrollRuntimeThr  >>> org.piwik.mobile2 <<<
    10-23 16:13:23.872 24532 24532 F DEBUG   : signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0x4df00024
    10-23 16:13:23.872 24532 24532 F DEBUG   :     r0 e00c3840  r1 e518a000  r2 00000003  r3 4df00000
    10-23 16:13:23.872 24532 24532 F DEBUG   :     r4 d282e4f8  r5 e00ec120  r6 d293f2cc  r7 00001630
    10-23 16:13:23.872 24532 24532 F DEBUG   :     r8 00000000  r9 d293f330  sl e00c3840  fp d293f334
    10-23 16:13:23.872 24532 24532 F DEBUG   :     ip d282e5d8  sp d293f2a0  lr d233c553  pc d241244c  cpsr 600b0010
    10-23 16:13:23.873 24532 24532 F DEBUG   : 
    10-23 16:13:23.873 24532 24532 F DEBUG   : backtrace:
    10-23 16:13:23.875 24532 24532 F DEBUG   :     #00 pc 0015744c  /data/app/org.piwik.mobile2-1/lib/arm/libkroll-v8.so (_ZN2v86Object3GetENS_6HandleINS_5ValueEEE+16)

**Comment by tsteur** [comment]
it's probably the same as in https://github.com/piwik/piwik-mobile-2/issues/5357 Problem is I cannot really reproduce it on the emulator and I don't have a device running Android 7.X

If I give you an updated APK, can you give it a try?

**Comment by Aeyoun** [comment]
> If I give you an updated APK, can you give it a try?

I can indeed.

**Comment by tsteur** [comment]
Sweet. Thank you! Link is here: https://piwik.org/wp-content/uploads/2016/10/PiwikMobile2.3.0-b1.apk

**Comment by tsteur** [comment]
Would you mind also trying to change the period to eg "week" in the date chooser? The emulator crashes when doing this on Android 7 but it works on my phone which runs on Android 6

**Comment by tsteur** [comment]
@Aeyoun have you had a chance to give it a try yet?

**Comment by Aeyoun** [comment]
I’m sorry, I only revived an email when you `@`mentioned me.  Not sure why as I should receive notifications when I created the issue! Following up on your feedback now. Let me see … .

* The linked `PiwikMobile2.3.0-b2.apk`  build crashes in the same way.
* Changing the period doesn’t crash the app, but the app does crash when I click the Home button and then reopen the app again.

**Comment by tsteur** [comment]
Are there any specific steps in order for it to crash or do you simple open it and then click the home button and open it again? At least on stock android I cannot reproduce it, which makes it a bit hard. By any chance, do you have maybe development mode and some certain settings enabled? @Aeyoun

**Comment by Aeyoun** [comment]
@tsteur, hm – not really. Sometimes it wouldn’t reproduce immediately but I don’t now what is difference when this happens. If I open two–three other apps after returning to the Home screen and before opening Piwik again, then it always reproduces.

**Comment by tsteur** [comment]
I have tried to reproduce with an emulator on Android 7.0 and 7.1.1 but can't :( I have forced garbage collection etc on the emulator but always worked for me. Are you maybe using any other apps like greenify or apps that do something memory or performance/battery boosting related etc?

We might release a new version 2.3.0 already next week but would then try to fix the bug in 2.3.1. I hope you understand it is a bit hard to look into when not being able to reproduce

**Comment by Aeyoun** [comment]
> Are you maybe using any other apps like greenify or apps that do something memory or performance/battery boosting related etc?

Nope, nothing like that,

Aren't the crashlogs submitted through Google Play any use?

**Comment by tsteur** [comment]
Not really. I'm only seeing only a very few reports there from Nexus 5 and they all only contain:

```
Backtrace:
#00 pc 0015744c  /data/app/org.piwik.mobile2-1/lib/arm/libkroll-v8.so (_ZN2v86Object3GetENS_6HandleINS_5ValueEEE+16)
```

Which is from an underlying SDK we are using to build the app. Can you maybe send me a new crash report then I can maybe see more or ask in a forum about the error but likely it'll be hard for them to say anything without my providing a reproduceable example

**Comment by Aeyoun** [comment]
> they all only contain libkroll-v8.so

That seems like what I’m submitting. Are all the logs coming from the same IP in Norway? 😉 

> Can you maybe send me a new crash report then I can maybe see more or ask in a forum about the error but likely it'll be hard for them to say anything without my providing a reproduceable example

So I wiped my phone clean (ouch), installed Piwik from the Play Store, opened it and logged in, clicked back to Home, and then clicked on the app icon. The app crashed straight away. Nexus 5X, submitted it through Google Play.

The weird thing is, it’s much harder to reproduce when plugged in to USB. I’m not sure what that means, except that the Battery Optimizer in Android 6+ isn’t enabled when the device is plugged in to a power supply. Also makes it harder to debug.

**Comment by tsteur** [comment]
Google is not showing me the report probably because the version has not been published on the Marketplace yet. I will try to see if I can do some tests with the battery optimizer later. On Android 6 it worked for me with the optimizer.

Unfortunately Google doesn't let me see the country :)

**Comment by tsteur** [comment]
@da2x  could you try another version? https://piwik.org/wp-content/uploads/2016/11/PiwikMobile2.3.0-b4.apk

It is using a newer SDK version and I have changed something in the code. To test can you open the app, and at least once open the left sidebar where it shows the list of available reports?  Then close the app. Maybe it will be closed correctly then.

**Comment by da2x** [comment]
Nope. It crashes just as often regardless of what I do with the left sidebar or not. This build doesn’t produce any output in logcat when it crashes, so I can’t say whether it’s the exact same problem or not.

**Comment by tsteur** [comment]
Thank you for testing again. We have released an update but it won't fix this issue yet. I have no clue how this happens and cannot reproduce it :( I will ask around to see if maybe someone has this device but they are not selling it here in New Zealand so changes are low. I might ask you soon again to test something if that's ok with you :)

**Comment by tsteur** [comment]
@da2x could you maybe give https://www.dropbox.com/s/3tm76ogfv4tte6p/Piwik%20Mobile%203.apk?dl=0 a try to see if this still happens?

**Comment by da2x** [comment]
(I’ll look into this early next week.)

**Comment by da2x** [comment]
W TiAndroid: (KrollRuntimeThread) [2599,3077] Application instance no longer available. Unable to get current activity.
    W TiAndroid: (KrollRuntimeThread) [3,3080] Application instance no longer available. Unable to get current activity.
    W TiAndroid: (KrollRuntimeThread) [1,3081] Application instance no longer available. Unable to get current activity.
    I TiRootActivity: (main) [0,0] checkpoint, on root activity create, savedInstanceState: Bundle[{android:viewHierarchyState=Bundle[mParcelledData.dataSize=528], android:lastAutofillId=1073741823}]
    I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=org.piwik.mobile2/.PiwikMobile2Activity (has extras)} from uid 10045
    W WindowManager: TaskSnapshotSurface.create: Failed to find main window for token=AppWindowToken{e1c2eb9 token=Token{e2ae880 ActivityRecord{b52e103 u0 org.piwik.mobile2/.PiwikMobile2Activity t5041}}}
    I ActivityManager: Process org.piwik.mobile2 (pid 4064) has died: cch  CAC 
    W InputDispatcher: channel 'c008001 org.piwik.mobile2/org.piwik.mobile2.PiwikMobile2Activity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
    E InputDispatcher: channel 'c008001 org.piwik.mobile2/org.piwik.mobile2.PiwikMobile2Activity (server)' ~ Channel is unrecoverably broken and will be disposed!
    W InputDispatcher: channel 'ff41b3d org.piwik.mobile2/org.appcelerator.titanium.TiActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
    E InputDispatcher: channel 'ff41b3d org.piwik.mobile2/org.appcelerator.titanium.TiActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
    I WindowManager: WIN DEATH: Window{c008001 u0 org.piwik.mobile2/org.piwik.mobile2.PiwikMobile2Activity}
    W InputDispatcher: Attempted to unregister already unregistered input channel 'c008001 org.piwik.mobile2/org.piwik.mobile2.PiwikMobile2Activity (server)'
    I WindowManager: WIN DEATH: Window{ff41b3d u0 org.piwik.mobile2/org.appcelerator.titanium.TiActivity}
    W InputDispatcher: Attempted to unregister already unregistered input channel 'ff41b3d org.piwik.mobile2/org.appcelerator.titanium.TiActivity (server)'
    I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=org.piwik.mobile2/.PiwikMobile2Activity (has extras)} from uid 10045
    I ActivityManager: Start proc 5188:org.piwik.mobile2/u0a178 for activity org.piwik.mobile2/.PiwikMobile2Activity
    I ActivityManager: START u0 {cmp=org.piwik.mobile2/org.appcelerator.titanium.TiActivity (has extras)} from uid 10178
    I TiRootActivity: (main) [0,0] checkpoint, on root activity resume. activity = org.piwik.mobile2.PiwikMobile2Activity@7178f06
    I ActivityManager: Displayed org.piwik.mobile2/org.appcelerator.titanium.TiActivity: +1s127ms (total +5s402ms)
    I LatinIme: onActivate() : EditorInfo = Package = org.piwik.mobile2 : Type = Text : Learning = Disable : Suggestion = Hide : AutoCorrection = Disable : Microphone = Show : Incognito = Disable

No signs of the oomreaper. Tried on three different phones: galaxy s8, nexus 5, nexus 5x

**Comment by tsteur** [comment]
Thank you @da2x I will see what I can do and hope to find more information based on the log

**Comment by tsteur** [comment]
Wonder if starting in foreground or something like this helps: https://jira.appcelerator.org/browse/TIMOB-3286

https://jira.appcelerator.org/browse/TIMOB-1911 might help as well
